gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java

```diff
@@ -924,6 +926,227 @@ public void deleteObjects(List<StorageResourceId> fullObjectNames) throws IOExce
     }
   }
 
+  /** See {@link GoogleCloudStorage#deleteFolders(List)} for details about expected behavior. */
+  @Override
+  public void deleteFolders(List<FolderInfo> folders) throws IOException {
+
+    if (folders.isEmpty()) {
+      return;
+    }
+
+    // Gather exceptions to wrap in a composite exception at the end.
+    final KeySetView<IOException, Boolean> innerExceptions = ConcurrentHashMap.newKeySet();
+    String traceContext = String.format("batchFolderDelete(size=%s)", folders.size());
+    try (ITraceOperation to = TraceOperation.addToExistingTrace(traceContext)) {
+
+      // threads for parallel delete of folder resources
+      BatchHelper batchHelper =
+          batchFactory.newBatchHelper(
+              httpRequestInitializer,
+              storage,
+              storageOptions.getMaxRequestsPerBatch(),
+              folders.size(),
+              storageOptions.getBatchThreads(),
+              "batchFolderDelete");
+
+      // Map to store number of children for each parent object
+      HashMap<String, Long> occurenceOfFolder = new HashMap<>();
+      for (FolderInfo currentFolder : folders) {
+        // inserting each folder
+        if (!occurenceOfFolder.containsKey(currentFolder)) {
```

Comment:
The key is string. How can we use FolderInfo object as key?

Reply:
Also, this always be true, correct?

Reply:
FolderInfo has FolderName which is essentially the path of the folder (excluding bucket) and that always has to be unique. I believe we cannot have two folder objects with same path.

It might be possible that when we were adding some folderresource A/B/, it's parent A/ was also added the map by the loop, now when loop is processing for A/ we dont want it to be processed.